astropy/io/ascii/qdp.py

```diff
@@ -0,0 +1,578 @@
+    _line_type_re = re.compile(_type_re)
+    line = line.strip()
+    if not line:
+        return 'comment'
+    match = _line_type_re.match(line)
+
+    if match is None:
+        raise ValueError(f'Unrecognized QDP line: {line}')
+    for type_, val in match.groupdict().items():
+        if val is None:
+            continue
+        if type_ == 'data':
+            return f'data,{len(val.split(sep=delimiter))}'
+        else:
+            return type_
+
+
+def _get_type_from_list_of_lines(lines, delimiter=None):
+    """Read through the list of QDP file lines and label each line by type
+
+    Parameters
+    ----------
+    lines : list
+        List containing one file line in each entry
+
+    Returns
+    -------
+    contents : list
+        List containing the type for each line (see `line_type_and_data`)
+    ncol : int
+        The number of columns in the data lines. Must be the same throughout
+        the file
+
+    Examples
+    --------
+    >>> line0 = "! A comment"
+    >>> line1 = "543 12 456.0"
+    >>> lines = [line0, line1]
+    >>> types, ncol = _get_type_from_list_of_lines(lines)
+    >>> types[0]
+    'comment'
+    >>> types[1]
+    'data,3'
+    >>> ncol
+    3
+    >>> lines.append("23")
+    >>> _get_type_from_list_of_lines(lines)
+    Traceback (most recent call last):
+        ...
+    ValueError: Inconsistent number of columns
+    """
+
+    types = [_line_type(line, delimiter=delimiter) for line in lines]
+    current_ncol = None
+    for type_ in types:
+        if type_.startswith('data', ):
+            ncol = int(type_[5:])
+            if current_ncol is None:
+                current_ncol = ncol
+            elif ncol != current_ncol:
+                raise ValueError('Inconsistent number of columns')
+
+    return types, current_ncol
+
+
+def _get_lines_from_file(qdp_file):
+    if "\n" in qdp_file:
+        lines = qdp_file.split("\n")
+    elif isinstance(qdp_file, str):
+        with open(qdp_file) as fobj:
+            lines = fobj.readlines()
+    elif isinstance(qdp_file, Iterable):
+        lines = qdp_file
+
+    return lines
+
+
+# def _analyze_qdp_file(qdp_file):
+#     """Read through the QDP file and label each line by type
+#
+#     Parameters
+#     ----------
+#     qdp_file : str
+#         File name
+#
+#     Returns
+#     -------
+#     contents : list
+#         List containing the type for each line (see `line_type_and_data`)
+#     ncol : int
+#         The number of columns in the data lines. Must be the same throughout
+#         the file
+#     """
+#     lines = _get_lines_from_file(qdp_file)
+#     return _get_type_from_list_of_lines(lines)
+
+
+def _interpret_err_lines(err_specs, ncols, names=None):
+    """Give list of column names from the READ SERR and TERR commands
+
+    Parameters
+    ----------
+    err_specs : dict, {'serr': [n0, n1, ...], 'terr': [n2, n3, ...]}
```

Comment:
The top line should just be `err_specs : dict` (i.e. only specify the type) with the details in the next line.

See https://github.com/astropy/astropy/pull/11118/files for examples of where this is coming from.